diff --git a/target/linux/kirkwood/ea3500-alt/config-default b/target/linux/kirkwood/ea3500-alt/config-default
index e69de29..9bb22cf 100644
--- a/target/linux/kirkwood/ea3500-alt/config-default
+++ b/target/linux/kirkwood/ea3500-alt/config-default
@@ -0,0 +1,3 @@
+CONFIG_CMDLINE="ubi.mtd=6 root=ubi0:rootfs rootfstype=ubifs noinitrd console=ttyS0,115200"
+CONFIG_CMDLINE_FORCE=y
+# CONFIG_CMDLINE_FROM_BOOTLOADER is not set
diff --git a/target/linux/kirkwood/ea3500-alt/files/etc/fw_env.config b/target/linux/kirkwood/ea3500-alt/files/etc/fw_env.config
index e69de29..1673d85 100644
--- a/target/linux/kirkwood/ea3500-alt/files/etc/fw_env.config
+++ b/target/linux/kirkwood/ea3500-alt/files/etc/fw_env.config
@@ -0,0 +1,2 @@
+# MTD device name	Device offset	Env. size	Flash sector size	Number of sectors
+/dev/mtd1		0x00000		0x20000		0x20000			1
diff --git a/target/linux/kirkwood/ea3500-alt/profiles/115-router.mk b/target/linux/kirkwood/ea3500-alt/profiles/115-router.mk
index e69de29..2047df5 100644
--- a/target/linux/kirkwood/ea3500-alt/profiles/115-router.mk
+++ b/target/linux/kirkwood/ea3500-alt/profiles/115-router.mk
@@ -0,0 +1,22 @@
+#
+# Copyright (C) 2014 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/EA3500
+  NAME:=Linksys EA3500
+  PACKAGES:= \
+	kmod-mwl8k kmod-usb2 kmod-usb-storage \
+	swconfig uboot-envtools
+endef
+
+define Profile/EA3500/Description
+ Package set compatible with Linksys EA3500 board.
+endef
+
+EA3500_UBIFS_OPTS:="-m 512 -e 15872 -c 4096"
+EA3500_UBI_OPTS:="-m 512 -p 16384 -s 256"
+
+$(eval $(call Profile,EA3500))
diff --git a/target/linux/kirkwood/ea3500-alt/target.mk b/target/linux/kirkwood/ea3500-alt/target.mk
index e69de29..3654e81 100644
--- a/target/linux/kirkwood/ea3500-alt/target.mk
+++ b/target/linux/kirkwood/ea3500-alt/target.mk
@@ -0,0 +1,12 @@
+#
+# Copyright (C) 2008-2010 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+BOARDNAME:=Linksys EA3500 (alt)
+
+define Target/Description
+	Build firmware images for Linksys EA3500 for use on the primary kernel / rootfs partitions (i.e. when the image that you'll be flashing from is on the 'alt' partitions).
+endef
diff --git a/target/linux/kirkwood/ea3500-pri/config-default b/target/linux/kirkwood/ea3500-pri/config-default
index e69de29..a3ac961 100644
--- a/target/linux/kirkwood/ea3500-pri/config-default
+++ b/target/linux/kirkwood/ea3500-pri/config-default
@@ -0,0 +1,3 @@
+CONFIG_CMDLINE="ubi.mtd=4 root=ubi0:rootfs rootfstype=ubifs noinitrd console=ttyS0,115200"
+CONFIG_CMDLINE_FORCE=y
+# CONFIG_CMDLINE_FROM_BOOTLOADER is not set
diff --git a/target/linux/kirkwood/ea3500-pri/files/etc/fw_env.config b/target/linux/kirkwood/ea3500-pri/files/etc/fw_env.config
index e69de29..1673d85 100644
--- a/target/linux/kirkwood/ea3500-pri/files/etc/fw_env.config
+++ b/target/linux/kirkwood/ea3500-pri/files/etc/fw_env.config
@@ -0,0 +1,2 @@
+# MTD device name	Device offset	Env. size	Flash sector size	Number of sectors
+/dev/mtd1		0x00000		0x20000		0x20000			1
diff --git a/target/linux/kirkwood/ea3500-pri/profiles/115-router.mk b/target/linux/kirkwood/ea3500-pri/profiles/115-router.mk
index e69de29..2047df5 100644
--- a/target/linux/kirkwood/ea3500-pri/profiles/115-router.mk
+++ b/target/linux/kirkwood/ea3500-pri/profiles/115-router.mk
@@ -0,0 +1,22 @@
+#
+# Copyright (C) 2014 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/EA3500
+  NAME:=Linksys EA3500
+  PACKAGES:= \
+	kmod-mwl8k kmod-usb2 kmod-usb-storage \
+	swconfig uboot-envtools
+endef
+
+define Profile/EA3500/Description
+ Package set compatible with Linksys EA3500 board.
+endef
+
+EA3500_UBIFS_OPTS:="-m 512 -e 15872 -c 4096"
+EA3500_UBI_OPTS:="-m 512 -p 16384 -s 256"
+
+$(eval $(call Profile,EA3500))
diff --git a/target/linux/kirkwood/ea3500-pri/target.mk b/target/linux/kirkwood/ea3500-pri/target.mk
index e69de29..7a8c2f3 100644
--- a/target/linux/kirkwood/ea3500-pri/target.mk
+++ b/target/linux/kirkwood/ea3500-pri/target.mk
@@ -0,0 +1,12 @@
+#
+# Copyright (C) 2008-2010 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+BOARDNAME:=Linksys EA3500 (pri)
+
+define Target/Description
+	Build firmware images for Linksys EA3500 for use on the primary kernel / rootfs partitions (i.e. when the image that you'll be flashing from is on the 'alt' partitions).
+endef
diff --git a/target/linux/kirkwood/ea4500-alt/config-default b/target/linux/kirkwood/ea4500-alt/config-default
index e69de29..9bb22cf 100644
--- a/target/linux/kirkwood/ea4500-alt/config-default
+++ b/target/linux/kirkwood/ea4500-alt/config-default
@@ -0,0 +1,3 @@
+CONFIG_CMDLINE="ubi.mtd=6 root=ubi0:rootfs rootfstype=ubifs noinitrd console=ttyS0,115200"
+CONFIG_CMDLINE_FORCE=y
+# CONFIG_CMDLINE_FROM_BOOTLOADER is not set
diff --git a/target/linux/kirkwood/ea4500-alt/files/etc/fw_env.config b/target/linux/kirkwood/ea4500-alt/files/etc/fw_env.config
index e69de29..1673d85 100644
--- a/target/linux/kirkwood/ea4500-alt/files/etc/fw_env.config
+++ b/target/linux/kirkwood/ea4500-alt/files/etc/fw_env.config
@@ -0,0 +1,2 @@
+# MTD device name	Device offset	Env. size	Flash sector size	Number of sectors
+/dev/mtd1		0x00000		0x20000		0x20000			1
diff --git a/target/linux/kirkwood/ea4500-alt/profiles/115-router.mk b/target/linux/kirkwood/ea4500-alt/profiles/115-router.mk
index e69de29..d768bbb 100644
--- a/target/linux/kirkwood/ea4500-alt/profiles/115-router.mk
+++ b/target/linux/kirkwood/ea4500-alt/profiles/115-router.mk
@@ -0,0 +1,22 @@
+#
+# Copyright (C) 2014 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/EA4500
+  NAME:=Linksys EA4500
+  PACKAGES:= \
+	kmod-mwl8k kmod-usb2 kmod-usb-storage \
+	swconfig uboot-envtools
+endef
+
+define Profile/EA4500/Description
+ Package set compatible with Linksys EA4500 board.
+endef
+
+EA4500_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+EA4500_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,EA4500))
diff --git a/target/linux/kirkwood/ea4500-alt/target.mk b/target/linux/kirkwood/ea4500-alt/target.mk
index e69de29..11085cb 100644
--- a/target/linux/kirkwood/ea4500-alt/target.mk
+++ b/target/linux/kirkwood/ea4500-alt/target.mk
@@ -0,0 +1,12 @@
+#
+# Copyright (C) 2008-2010 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+BOARDNAME:=Linksys EA4500 (alt)
+
+define Target/Description
+	Build firmware images for Linksys EA4500 / E4200v2 for use on the 'alt' kernel / rootfs partitions (i.e. when the image that you'll be flashing from is on the primary partitions).
+endef
diff --git a/target/linux/kirkwood/ea4500-pri/config-default b/target/linux/kirkwood/ea4500-pri/config-default
index e69de29..a3ac961 100644
--- a/target/linux/kirkwood/ea4500-pri/config-default
+++ b/target/linux/kirkwood/ea4500-pri/config-default
@@ -0,0 +1,3 @@
+CONFIG_CMDLINE="ubi.mtd=4 root=ubi0:rootfs rootfstype=ubifs noinitrd console=ttyS0,115200"
+CONFIG_CMDLINE_FORCE=y
+# CONFIG_CMDLINE_FROM_BOOTLOADER is not set
diff --git a/target/linux/kirkwood/ea4500-pri/files/etc/fw_env.config b/target/linux/kirkwood/ea4500-pri/files/etc/fw_env.config
index e69de29..1673d85 100644
--- a/target/linux/kirkwood/ea4500-pri/files/etc/fw_env.config
+++ b/target/linux/kirkwood/ea4500-pri/files/etc/fw_env.config
@@ -0,0 +1,2 @@
+# MTD device name	Device offset	Env. size	Flash sector size	Number of sectors
+/dev/mtd1		0x00000		0x20000		0x20000			1
diff --git a/target/linux/kirkwood/ea4500-pri/profiles/115-router.mk b/target/linux/kirkwood/ea4500-pri/profiles/115-router.mk
index e69de29..d768bbb 100644
--- a/target/linux/kirkwood/ea4500-pri/profiles/115-router.mk
+++ b/target/linux/kirkwood/ea4500-pri/profiles/115-router.mk
@@ -0,0 +1,22 @@
+#
+# Copyright (C) 2014 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/EA4500
+  NAME:=Linksys EA4500
+  PACKAGES:= \
+	kmod-mwl8k kmod-usb2 kmod-usb-storage \
+	swconfig uboot-envtools
+endef
+
+define Profile/EA4500/Description
+ Package set compatible with Linksys EA4500 board.
+endef
+
+EA4500_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+EA4500_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,EA4500))
diff --git a/target/linux/kirkwood/ea4500-pri/target.mk b/target/linux/kirkwood/ea4500-pri/target.mk
index e69de29..902861b 100644
--- a/target/linux/kirkwood/ea4500-pri/target.mk
+++ b/target/linux/kirkwood/ea4500-pri/target.mk
@@ -0,0 +1,12 @@
+#
+# Copyright (C) 2008-2010 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+BOARDNAME:=Linksys EA4500 (pri)
+
+define Target/Description
+	Build firmware images for Linksys EA4500 / E4200v2 for use on the primary kernel / rootfs partitions (i.e. when the image that you'll be flashing from is on the 'alt' partitions).
+endef
diff --git a/target/linux/kirkwood/generic/profiles/100-generic.mk b/target/linux/kirkwood/generic/profiles/100-generic.mk
index e69de29..da45604 100644
--- a/target/linux/kirkwood/generic/profiles/100-generic.mk
+++ b/target/linux/kirkwood/generic/profiles/100-generic.mk
@@ -0,0 +1,21 @@
+#
+# Copyright (C) 2013 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/Generic
+  NAME:=Generic (default)
+  PACKAGES:= \
+	kmod-mmc kmod-mvsdio kmod-usb2 kmod-usb-storage \
+	kmod-of-i2c kmod-i2c-core kmod-i2c-mv64xxx \
+	kmod-ata-core kmod-ata-marvell-sata \
+	kmod-rtc-marvell kmod-thermal-kirkwood
+endef
+
+define Profile/Generic/Description
+ Package set compatible with most Marvell Kirkwood based boards.
+endef
+
+$(eval $(call Profile,Generic))
diff --git a/target/linux/kirkwood/generic/profiles/110-nas.mk b/target/linux/kirkwood/generic/profiles/110-nas.mk
index e69de29..66a4f6c 100644
--- a/target/linux/kirkwood/generic/profiles/110-nas.mk
+++ b/target/linux/kirkwood/generic/profiles/110-nas.mk
@@ -0,0 +1,103 @@
+#
+# Copyright (C) 2013 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/DOCKSTAR
+  NAME:=Seagate FreeAgent Dockstar
+  PACKAGES:= \
+	kmod-rtc-marvell kmod-usb2 kmod-usb-storage \
+	uboot-envtools
+endef
+
+define Profile/DOCKSTAR/Description
+ Package set compatible with Seagate FreeAgent Dockstar board.
+endef
+
+DOCKSTAR_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+DOCKSTAR_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,DOCKSTAR))
+
+define Profile/GOFLEXNET
+  NAME:=Seagate GoFlexNet
+  PACKAGES:= \
+	kmod-ata-core kmod-ata-marvell-sata \
+	kmod-usb2 kmod-usb-storage \
+	uboot-envtools
+endef
+
+define Profile/GOFLEXNET/Description
+ Package set compatible with Seagate GoFlexNet
+endef
+
+GOFLEXNET_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+GOFLEXNET_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,GOFLEXNET))
+
+define Profile/IB62X0
+  NAME:=RaidSonic ICY BOX IB-NAS62x0
+  PACKAGES:= \
+	kmod-ata-core kmod-ata-marvell-sata \
+	kmod-rtc-marvell kmod-usb2 kmod-usb-storage \
+	uboot-envtools
+endef
+
+define Profile/IB62X0/Description
+ Package set compatible with RaidSonic ICY BOX IB-NAS62x0 board.
+endef
+
+IB62X0_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+IB62X0_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,IB62X0))
+
+define Profile/ICONNECT
+  NAME:=Iomega Iconnect
+  PACKAGES:= \
+	kmod-usb2 kmod-usb-storage
+endef
+
+define Profile/ICONNECT/Description
+ Package set compatible with Iomega Iconnect board.
+endef
+
+ICONNECT_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+ICONNECT_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,ICONNECT))
+
+define Profile/IOMEGA_IX2_200
+  NAME:=Iomega StorCenter ix2-200
+  PACKAGES:= \
+	kmod-ata-core kmod-ata-marvell-sata \
+	kmod-usb2 kmod-usb-storage \
+	uboot-envtools
+endef
+
+define Profile/IOMEGA_IX2_200/Description
+ Package set compatible with Iomega StorCenter ix2-200 board.
+endef
+
+IOMEGA_IX2_200_UBIFS_OPTS:="-m 512 -e 15872 -c 2048"
+IOMEGA_IX2_200_UBI_OPTS:="-m 512 -p 16384 -s 256"
+
+$(eval $(call Profile,IOMEGA_IX2_200))
+
+define Profile/POGOE02
+  NAME:=Cloud Engines Pogoplug E02
+  PACKAGES:= \
+	kmod-usb2 kmod-usb-storage uboot-envtools
+endef
+
+define Profile/POGOE02/Description
+ Package set compatible with Cloud Engines Pogoplug E02 board.
+endef
+
+POGOE02_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+POGOE02_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,POGOE02))
diff --git a/target/linux/kirkwood/generic/profiles/120-plug.mk b/target/linux/kirkwood/generic/profiles/120-plug.mk
index e69de29..1879683 100644
--- a/target/linux/kirkwood/generic/profiles/120-plug.mk
+++ b/target/linux/kirkwood/generic/profiles/120-plug.mk
@@ -0,0 +1,78 @@
+#
+# Copyright (C) 2014 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/SHEEVAPLUG
+  NAME:=Globalscale Technologies SheevaPlug
+  PACKAGES:= \
+	kmod-mmc kmod-mvsdio kmod-usb2 kmod-usb-storage \
+	kmod-of-i2c kmod-i2c-core kmod-i2c-mv64xxx \
+	kmod-ata-core uboot-envtools
+endef
+
+define Profile/SHEEVAPLUG/Description
+ Package set compatible with Globalscale Technologies SheevaPlug board.
+endef
+
+SHEEVAPLUG_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+SHEEVAPLUG_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,SHEEVAPLUG))
+
+define Profile/SHEEVAPLUGSATA
+  NAME:=Globalscale Technologies eSATA SheevaPlug
+  PACKAGES:= \
+	kmod-mmc kmod-mvsdio kmod-usb2 kmod-usb-storage \
+	kmod-of-i2c kmod-i2c-core kmod-i2c-mv64xxx \
+	kmod-ata-core kmod-ata-marvell-sata \
+	uboot-envtools
+endef
+
+define Profile/SHEEVAPLUGSATA/Description
+ Package set compatible with Globalscale Technologies eSATA SheevaPlug board.
+endef
+
+SHEEVAPLUGSATA_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+SHEEVAPLUGSATA_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,SHEEVAPLUGSATA))
+
+define Profile/GuruplugServerPlus
+  NAME:=Globalscale Technologies Guruplug Server Plus
+  PACKAGES:= \
+	kmod-mmc kmod-mvsdio kmod-usb2 kmod-usb-storage \
+	kmod-of-i2c kmod-i2c-core kmod-i2c-mv64xxx \
+	kmod-ata-core kmod-ata-marvell-sata \
+	kmod-btmrvl kmod-btmrvl-sdio kmod-libertas kmod-libertas-sdio \
+	uboot-envtools wpad-mini
+endef
+
+define Profile/GuruplugServerPlus/Description
+ Package set compatible with Globalscale Technologies Guruplug Server Plus board.
+endef
+
+GuruplugServerPlus_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+GuruplugServerPlus_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,GuruplugServerPlus))
+
+define Profile/Topkick1281P2
+  NAME:=Univeral Scientific Industrial Co. Topkick-1281P2
+  PACKAGES:= \
+	kmod-mmc kmod-mvsdio kmod-usb2 kmod-usb-storage \
+	kmod-of-i2c kmod-i2c-core kmod-i2c-mv64xxx \
+	kmod-ata-core kmod-ata-marvell-sata \
+	uboot-envtools
+endef
+
+define Profile/Topkick1281P2/Description
+ Package set compatible with Univeral Scientific Industrial Co. Topkick-1281P2 board.
+endef
+
+Topkick1281P2_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
+Topkick1281P2_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
+
+$(eval $(call Profile,Topkick1281P2))
diff --git a/target/linux/kirkwood/generic/target.mk b/target/linux/kirkwood/generic/target.mk
index e69de29..b40eb1d 100644
--- a/target/linux/kirkwood/generic/target.mk
+++ b/target/linux/kirkwood/generic/target.mk
@@ -0,0 +1,12 @@
+#
+# Copyright (C) 2008-2010 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+BOARDNAME:=Generic
+
+define Target/Description
+	Build firmware images for Marvell Kirkwood based devices other than Linksys EA3500, E4200v2, EA4500.
+endef
diff --git a/target/linux/kirkwood/image/Makefile b/target/linux/kirkwood/image/Makefile
index ede7cb3..cf39ccb 100644
--- a/target/linux/kirkwood/image/Makefile
+++ b/target/linux/kirkwood/image/Makefile
@@ -9,6 +9,9 @@ include $(INCLUDE_DIR)/image.mk
 
 NAND_BLOCKSIZE := 2048-128k
 
+# Linksys EA4500 Kernel Partition is defined as 2M
+EA4500_KERNEL_MTD_SIZE := 2097152
+
 define sanitize_profile_name
 $(shell echo $(PROFILE) | tr '[:upper:]' '[:lower:]')
 endef
@@ -104,11 +107,34 @@ define Image/Build/ubifs
  endif
 endef
 
+define Image/Build/ssa
+
+	if [ `stat -c %s $(BIN_DIR)/$(IMG_PREFIX)-$(call sanitize_profile_name)-uImage` -gt $(EA4500_KERNEL_MTD_SIZE) ] ; then \
+		echo "$(BIN_DIR)/$(IMG_PREFIX)-$(call sanitize_profile_name)-uImage \>2M" >&2; \
+		exit 1; \
+	else \
+		( \
+			dd if=$(BIN_DIR)/$(IMG_PREFIX)-$(call sanitize_profile_name)-uImage bs=$(EA4500_KERNEL_MTD_SIZE) conv=sync; \
+			dd if=$(BIN_DIR)/$(IMG_PREFIX)-$(call sanitize_profile_name)-rootfs.ubi; \
+		) > $(BIN_DIR)/$(IMG_PREFIX)-$(call sanitize_profile_name).ssa; \
+	fi
+
+endef
+
 define Image/Build/ubi
 
  ifneq ($($(PROFILE)_UBI_OPTS),)
 	$(CP) $(KDIR)/root.ubi $(BIN_DIR)/$(IMG_PREFIX)-$(call sanitize_profile_name)-rootfs.ubi
  endif
+
+ ifeq ($(PROFILE),EA4500)
+	$(call Image/Build/ssa)
+ endif
+
+ ifeq ($(PROFILE),EA3500)
+	$(call Image/Build/ssa)
+ endif
+
 endef
 
 Image/BuildKernel/Template/Generic=$(call Image/BuildKernel/Template)
diff --git a/target/linux/kirkwood/patches-3.18/160-ea4500.patch b/target/linux/kirkwood/patches-3.18/160-ea4500.patch
index 386b5ea..161341f 100644
--- a/target/linux/kirkwood/patches-3.18/160-ea4500.patch
+++ b/target/linux/kirkwood/patches-3.18/160-ea4500.patch
@@ -10,7 +10,7 @@
  	kirkwood-ib62x0.dtb \
 --- /dev/null
 +++ b/arch/arm/boot/dts/kirkwood-ea4500.dts
-@@ -0,0 +1,167 @@
+@@ -0,0 +1,187 @@
 +/*
 + * kirkwood-ea4500.dts - Device Tree file for Linksys EA4500
 + *
@@ -144,9 +144,14 @@
 +	};
 +
 +	partition@80000 {
-+		label = "u-boot environment";
++		label = "u_env";
 +		reg = <0x80000 0x20000>;
 +	};
++	
++	partition@A0000 {
++		label = "s_env";
++		reg = <0xA0000 0x20000>;
++	};
 +
 +	partition@200000 {
 +		label = "kernel";
@@ -154,8 +159,23 @@
 +	};
 +
 +	partition@400000 {
-+		label = "root";
-+		reg = <0x400000 0x1c00000>;
++		label = "rootfs";
++		reg = <0x400000 0x1800000>;
++	};
++
++	partition@1C00000 {
++		label = "alt_kernel";
++		reg = <0x1C00000 0x200000>;
++	};
++
++	partition@1E00000 {
++		label = "alt_rootfs";
++		reg = <0x1E00000 0x1800000>;
++	};
++
++	partition@3600000 {
++		label = "syscfg";
++		reg = <0x3600000 0x4A00000>;
 +	};
 +};
 +
diff --git a/target/linux/kirkwood/patches-3.18/170-ea3500.patch b/target/linux/kirkwood/patches-3.18/170-ea3500.patch
index 84645de..25a2068 100644
--- a/target/linux/kirkwood/patches-3.18/170-ea3500.patch
+++ b/target/linux/kirkwood/patches-3.18/170-ea3500.patch
@@ -10,7 +10,7 @@
  	kirkwood-guruplug-server-plus.dtb \
 --- /dev/null
 +++ b/arch/arm/boot/dts/kirkwood-ea3500.dts
-@@ -0,0 +1,160 @@
+@@ -0,0 +1,180 @@
 +/*
 + * kirkwood-ea3500.dts - Device Tree file for Linksys EA3500
 + *
@@ -137,19 +137,39 @@
 +	};
 +
 +	partition@80000 {
-+		label = "u-boot environment";
++		label = "u_env";
 +		reg = <0x80000 0x20000>;
 +	};
 +
++	partition@A0000 {
++	   	label = "s_env";
++		reg = <0xA0000 0x20000>;
++	};
++
 +	partition@200000 {
 +		label = "kernel";
 +		reg = <0x200000 0x200000>;
 +	};
 +
 +	partition@400000 {
-+		label = "root";
++		label = "rootfs";
 +		reg = <0x400000 0x1200000>;
 +	};
++
++	partition@1600000 {
++		label = "alt_kernel";
++		reg = <0x1600000 0x200000>;
++	};
++
++	partition@1800000 {
++		label = "alt_rootfs";
++		reg = <0x1800000 0x1200000>;
++	};
++
++	partition@2a00000 {
++	   	label = "sysconfig";
++		reg = <0x2a00000 0x1600000>;
++	};
 +};
 +
 +&mdio {
diff --git a/target/linux/kirkwood/profiles/110-nas.mk b/target/linux/kirkwood/profiles/110-nas.mk
deleted file mode 100644
index 66a4f6c..0000000
--- a/target/linux/kirkwood/profiles/110-nas.mk
+++ /dev/null
@@ -1,103 +0,0 @@
-#
-# Copyright (C) 2013 OpenWrt.org
-#
-# This is free software, licensed under the GNU General Public License v2.
-# See /LICENSE for more information.
-#
-
-define Profile/DOCKSTAR
-  NAME:=Seagate FreeAgent Dockstar
-  PACKAGES:= \
-	kmod-rtc-marvell kmod-usb2 kmod-usb-storage \
-	uboot-envtools
-endef
-
-define Profile/DOCKSTAR/Description
- Package set compatible with Seagate FreeAgent Dockstar board.
-endef
-
-DOCKSTAR_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
-DOCKSTAR_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
-
-$(eval $(call Profile,DOCKSTAR))
-
-define Profile/GOFLEXNET
-  NAME:=Seagate GoFlexNet
-  PACKAGES:= \
-	kmod-ata-core kmod-ata-marvell-sata \
-	kmod-usb2 kmod-usb-storage \
-	uboot-envtools
-endef
-
-define Profile/GOFLEXNET/Description
- Package set compatible with Seagate GoFlexNet
-endef
-
-GOFLEXNET_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
-GOFLEXNET_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
-
-$(eval $(call Profile,GOFLEXNET))
-
-define Profile/IB62X0
-  NAME:=RaidSonic ICY BOX IB-NAS62x0
-  PACKAGES:= \
-	kmod-ata-core kmod-ata-marvell-sata \
-	kmod-rtc-marvell kmod-usb2 kmod-usb-storage \
-	uboot-envtools
-endef
-
-define Profile/IB62X0/Description
- Package set compatible with RaidSonic ICY BOX IB-NAS62x0 board.
-endef
-
-IB62X0_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
-IB62X0_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
-
-$(eval $(call Profile,IB62X0))
-
-define Profile/ICONNECT
-  NAME:=Iomega Iconnect
-  PACKAGES:= \
-	kmod-usb2 kmod-usb-storage
-endef
-
-define Profile/ICONNECT/Description
- Package set compatible with Iomega Iconnect board.
-endef
-
-ICONNECT_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
-ICONNECT_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
-
-$(eval $(call Profile,ICONNECT))
-
-define Profile/IOMEGA_IX2_200
-  NAME:=Iomega StorCenter ix2-200
-  PACKAGES:= \
-	kmod-ata-core kmod-ata-marvell-sata \
-	kmod-usb2 kmod-usb-storage \
-	uboot-envtools
-endef
-
-define Profile/IOMEGA_IX2_200/Description
- Package set compatible with Iomega StorCenter ix2-200 board.
-endef
-
-IOMEGA_IX2_200_UBIFS_OPTS:="-m 512 -e 15872 -c 2048"
-IOMEGA_IX2_200_UBI_OPTS:="-m 512 -p 16384 -s 256"
-
-$(eval $(call Profile,IOMEGA_IX2_200))
-
-define Profile/POGOE02
-  NAME:=Cloud Engines Pogoplug E02
-  PACKAGES:= \
-	kmod-usb2 kmod-usb-storage uboot-envtools
-endef
-
-define Profile/POGOE02/Description
- Package set compatible with Cloud Engines Pogoplug E02 board.
-endef
-
-POGOE02_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
-POGOE02_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
-
-$(eval $(call Profile,POGOE02))
diff --git a/target/linux/kirkwood/profiles/115-router.mk b/target/linux/kirkwood/profiles/115-router.mk
deleted file mode 100644
index 769284f..0000000
--- a/target/linux/kirkwood/profiles/115-router.mk
+++ /dev/null
@@ -1,38 +0,0 @@
-#
-# Copyright (C) 2014 OpenWrt.org
-#
-# This is free software, licensed under the GNU General Public License v2.
-# See /LICENSE for more information.
-#
-
-define Profile/EA3500
-  NAME:=Linksys EA3500
-  PACKAGES:= \
-	kmod-mwl8k kmod-usb2 kmod-usb-storage \
-	swconfig uboot-envtools
-endef
-
-define Profile/EA3500/Description
- Package set compatible with Linksys EA3500 board.
-endef
-
-EA3500_UBIFS_OPTS:="-m 512 -e 15872 -c 4096"
-EA3500_UBI_OPTS:="-m 512 -p 16384 -s 256"
-
-$(eval $(call Profile,EA3500))
-
-define Profile/EA4500
-  NAME:=Linksys EA4500
-  PACKAGES:= \
-	kmod-mwl8k kmod-usb2 kmod-usb-storage \
-	swconfig uboot-envtools
-endef
-
-define Profile/EA4500/Description
- Package set compatible with Linksys EA4500 board.
-endef
-
-EA4500_UBIFS_OPTS:="-m 2048 -e 126KiB -c 4096"
-EA4500_UBI_OPTS:="-m 2048 -p 128KiB -s 512"
-
-$(eval $(call Profile,EA4500))
